[FLINK-23202][rpc] Fail rpc requests for unreachable recipients eagerly

tillrohrmann opened the pull request:
This PR adds support for failing RPC result futures eagerly if the recipient turns out to be unreachable (e.g. terminated or remote ActorSystem is unreachable). This allows to react to failure situations earlier.

The PR contains the following main commits:

#### 74dc9a1: Introduce marker interface to filter out Flink related RPC messages

The marker interface rpc.messages.Message identifies all Flink related RPC messages. This can
be used to filter for these messages and to trigger specific actions.

#### da79d62: Fail rpc requests for unreachable recipients eagerly

This commit adds the DeadLettersActor that monitor Akka's dead letter mailbox. Whenever a
Flink Message is received if sends a failure to the sender to notify the sender about the
unreachability of the recipient. That way it is possible to eagerly fail the result futures,
and not having to wait on the rpc timeout.

#### fe6bdb0: Fail with AkkaRecipientUnreachableException if local actor has terminated

This commit replaces the AskTimeoutException with an AkkaRecipientUnreachableException if the local actor
has already terminated. This works by looking at the exception message. This ensures that we respond with
the right exception if the local actor is unreachable.

zentol reviewed (commented):
~~Potentially~~ Related test failure (also fails in backports):
```
Jul 01 14:33:34 [ERROR] testDelayedRegisterTaskExecutor(org.apache.flink.runtime.resourcemanager.ResourceManagerTaskExecutorTest)  Time elapsed: 0.113 s  <<< FAILURE!
Jul 01 14:33:34 java.lang.AssertionError: 
Jul 01 14:33:34 
Jul 01 14:33:34 Expected: a string containing "ResourceManagerGateway.registerTaskExecutor"
```

The backports also show a failure in [TaskManagerProcessFailureBatchRecoveryITCase#AbstractTaskManagerProcessFailureRecoveryTest.testTaskManagerProcessFailure](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=19786&view=logs&j=5c8e7682-d68f-54d1-16a2-a09310218a49&t=f508e270-48d6-5f1e-3138-42a17e0714f0); the job is failing because the task submission is not possible (probably when the TM has failed).

tillrohrmann commented on flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/DeadLettersActor.java:
```diff
@@ -0,0 +1,70 @@
+        if (deadLetter.message() instanceof Message) {
+            if (deadLetter.sender().equals(getContext().getSystem().deadLetters())) {
+                LOG.debug(
+                        "Could not deliver message {} with no sender to recipient {}. "
+                                + "This indicates that the actor terminated unexpectedly.",
```
If a message has no sender, then it does not expect a result (e.g. when using `ActorRef.tell`).

tillrohrmann commented on flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/exceptions/AkkaRecipientUnreachableException.java:
```diff
@@ -0,0 +1,31 @@
+ * Exception which indicates that the specified recipient is unreachable. This can either mean that
+ * the recipient has been terminated or that the remote ActorSystem is currently not reachable.
+ */
+public class AkkaRecipientUnreachableException extends RecipientUnreachableException {
```
It is only there for reasons of symmetry with the other existing `Akka*Exceptions`. I think it does not add value and I will remove it.

tillrohrmann commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/messages/LocalRpcInvocation.java:
```diff
@@ -69,7 +69,10 @@ public String toString() {
                 }
             }
 
-            toString = "LocalRpcInvocation(" + methodName + '(' + paramTypeStringBuilder + "))";
```
Indeed. Good catch. Thanks.

tillrohrmann commented on flink-rpc/flink-rpc-akka/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcServiceTest.java:
```diff
@@ -361,6 +363,30 @@ public void testAkkaRpcServiceShutDownWithFailingRpcEndpoints() throws Exception
     }
 
+    @Test
+    public void failsRpcResultImmediatelyIfEndpointIsStopped() throws Exception {
```
The code looks identical but both tests test different things: `AkkaRpcServiceTest#failsRpcResultImmediatelyIfEndpointIsStopped` tests that the `RecipientUnreachableException` is thrown if the endpoint is in the same `ActorSystem`. In contrast to that, `RemoteAkkaRpcActorTest#failsRpcResultImmediatelyIfEndpointIsStopped` tests that the same behaviour occurs if the endpoint is in a different `ActorSystem` (e.g. remote communication is needed). In fact the former case is why I had to change `AkkaInvocationHandler#resolveTimeoutException` because it only surfaced as a `AskTimeoutException` with a special message.

tillrohrmann commented:
Thanks for the review @zentol. I have addressed your comments and hardened the failing test cases.

The problem of the `ResourceManagerTaskExecutorTest.testDelayedRegisterTaskExecutor` was that I changed how the `TimeoutException` message is generated in the `AkkaInvocationHandler#resolveTimeoutException` which missed the class name of the RPC. I've changed this by adding the declaring class name to the `RpcInvocation` message and their `toString` method. I think this is a good improvement because it makes it clearer which class one is calling a RPC on.

The problem of the `TaskManagerProcessFailureBatchRecoveryITCase#AbstractTaskManagerProcessFailureRecoveryTest.testTaskManagerProcessFailure` is actually quite interesting. The test checks that a job gets recovered if a TM dies. The problem was that it was build on some very specific Flink behaviour to realize this. The test assumed that a recovery takes at least as long as the `akka.ask.timeout` because cancelling the tasks on the failed TM won't complete until this timeout expired. Moreover, it assumes that a TM will only be marked as dead if the hearbeat timeout expires. Therefore, by configuring the `akka.ask.timeout` to `100 s` and setting the `heartbeat.timeout` to `10 s`, it knew that there will only be 2 restart attempts because the second attempt will only be started after the `heartbeat.timeout` has been triggered.

With my change, this behaviour changes because the cancel call won't block until the `akka.ask.timeout`. Instead it completes (or rather fails) right away, because it sees that the remote endpoint is no longer reachable. Therefore (and because we have set the delay between restarts to `0`), we will start right away with the next execution attempt. Since the dead TM has not been removed from the slot pool, we will reuse the slots for the second attempt and fail right away again when deploying the tasks. Since the TM will only be marked as dead after `10 s`, we will deplete our restart attempts and then fail the job.

The solution to fixing this problem is to set the number of restart attempts to `Integer.MAX_VALUE` and to configure a delay between restarts of `1 s` so that we don't have a busy loop. With FLINK-23209, this will improve considerably, because then it will only take the `heartbeat.interval` to detect the dead TM.

tillrohrmann commented:
However, this also creates a question: Do we want to backport this improvement to the `release-1.12.` and `release-1.13` branch? The PR changes slightly the behaviour of Flink in the sense that `akka.ask.timeout` can no longer be used to delay the restart attempts. Hence, when backporting this feature we might break some setups in which users implicitly relied on this behaviour.

zentol commented:
> akka.ask.timeout can no longer be used to delay the restart attempts

I'm not too concerned about this; we've had proper APIs to control restart delays for a long time.

> Do we want to backport this improvement to the release-1.12. and release-1.13 branch? 

Have we fully grasped the implications of this change yet?
In the case of a truly unreachable TM, these changes (along with FLINK-23209) are a clear improvement, but what about cases of intermittent networking issues?
Can the tiniest networking issue cause a TM to be considered unreachable and cause all tasks to be re-deployed? The heatbeat interval+timeout safeguarded against such issues because in practice multiple heartbeats had to fail.

tillrohrmann commented:
It is true that with #16357, intermittent networking issues will lead to more disconnections and consequently more re-deployments of tasks. Maybe this is something we want to make configurable (e.g. also having a threshold but weighing the importance of this signal a bit higher than when the last heartbeat was received). The problem with intermittent networking issues is always that you don't know how long they last.

We probably should have this discussion on the respective #16357 PR.

zentol reviewed (approved):
+1 to merge to master.

tillrohrmann commented:
Thanks for reviewing this PR @zentol. Merging this PR now.
